internal/cloud/azure/application_permission.go

```diff
@@ -160,15 +148,63 @@ func UpdateApplicationPermission(ctx context.Context, config *ApplicationPermiss
 
 func DeleteApplicationPermission(ctx context.Context, config *ApplicationPermissionConfig, raClient RoleAssignmentsClient, rdClient RoleDefinitionsClient) error {
 
-	// input := iam.DetachRolePolicyInput{
-	// 	RoleName:  &config.RoleName,
-	// 	PolicyArn: &config.PolicyARN,
-	// }
+	id, err := parseRoleAssignmentId(config.ID)
+	if err != nil {
+		return err
+	}
 
-	// _, deleteErr := client.DetachRolePolicy(ctx, &input)
-	// if deleteErr != nil {
-	// 	return deleteErr
-	// }
+	_, err = raClient.Delete(ctx, id.scope, id.name, "") //id.tenantId)
+	if err != nil {
+		return fmt.Errorf("deletion of Role Assignment %q returned an error: %w", config.ID, err)
+	}
 
 	return nil
 }
+
+type roleAssignmentId struct {
+	scope    string
+	name     string
+	tenantId string
+}
+
+func parseRoleAssignmentId(input string) (*roleAssignmentId, error) {
+	segments := strings.Split(input, "/providers/Microsoft.Authorization/roleAssignments/")
+	if len(segments) != 2 {
+		return nil, fmt.Errorf("expected Role Assignment ID to be in the format `{scope}/providers/Microsoft.Authorization/roleAssignments/{name}` but got %q", input)
+	}
+
+	// /{scope}/providers/Microsoft.Authorization/roleAssignments/{roleAssignmentName}
+	// Tenant ID only required when going cross-tenant
+	id := roleAssignmentId{
+		scope:    strings.TrimPrefix(segments[0], "/"),
+		name:     segments[1],
+		tenantId: "",
+	}
+	return &id, nil
+}
+
+func createRoleAssignment(ctx context.Context, scope string, roleAssignmentName string, parameters authorization.RoleAssignmentCreateParameters, raClient RoleAssignmentsClient) (authorization.RoleAssignment, error) {
```

Comment:
Don't rack your brain too hard here, again its [mostly copied](https://github.com/hashicorp/terraform-provider-azurerm/blob/accdeebe6072390c531b88f8e322b80acceecfb8/internal/services/authorization/role_assignment_resource.go#L317-L327) with our own retry wrapper.

The reason for this is that the service principal `create` can return very quickly, but that doesn't mean you can attach roles yet.  I tried backing this "polling" back to the SP creation itself, but of course that doesn't work...reads and lookups work just fine.  Just these role attachments (and who knows what else) will fail until after some arbitrary time when its "ready".

Azure does the same thing in their provider...the polling occurs downstream instead of the SP creation blocking until its "ready". Its probably just not easy to determine via the API.